### ISSUE TITLE
Prefer defined?(Namespace::VERSION) over defined?(Namespace)

### ISSUE THREAD

**Issue by nattfodd** [rubocop-hq/rubocop#5474]
**TLDR:** Some dependencies in some gems might have `Rails` module defined without `rails` gem itself, therefore `defined?(Rails)` check isn't correct and leads to misbeharior.

Some libraries which have a part of some Rails-specific logic included, usually check presence of Rails using `defined?(Rails)`. That causes troubles in certain situations - see https://github.com/toptal/chewy/pull/625

The idea may be expanded outside of Rails scope (i.e. checking Namespace::VERSION instead of just Namespace), but it requires more discussion as I'm not sure if it's applicable to every case.

Something like:

```
# Prefer using defined?(Namespace::VERSION) over defined?(Namespace)
Lint/LibDefinitionCheck:
  Enabled: true
```

**Comment by shevegen** [comment]
For the former, Namespace.const_defined? :VERSION might also work. Not sure how common it is - while I guess many projects use a VERSION constant (I do in my projects for example), I don't know how many other projects do.

I guess the underlying problem is that people in the gems mentioned in your link (aka rails-html-sanitizer and rails-dom-testing) assume that there is only one project called Rails ever ("one ring to bind them"), whereas it could be any other namespace defined by people. I ran into something somewhat similar with the "Configuration" namespace, and the configuration gem on rubygems.org.

In the long run, perhaps "namespaces" in ruby could also obtain additional identifiers for that namespace ... something like meta-fields, like based on the main author or so ("if constant is rails and if author is DHH, do that") - but I digress.

**Comment by Drenmi** [comment]
> For the former, `Namespace.const_defined? :VERSION` might also work.

This likely won't work so well, since the initial implementation is there to check if `Namespace` is defined. If it isn't, there isn't anything there to respond to `#const_defined?`, so you'll end up with a `NameError`. 🙂

**Comment by bbatsov** [comment]
There hasn't been much activity on this ticket and our Core Team is spread too thin on so many tasks, so I'll just close it for the sake of having a cleaner lists of tasks to focus on. We'd gladly review a PR, but it's unlikely that'd we're going to tackle this ourselves in the foreseeable future.
